main/static/script.js

```diff
@@ -94,107 +74,98 @@ module.controller("mainCtrl", function(
       }));
     }
   });
+})
 
-  // Triggers when the button is clicked.
-  $scope.onSubmitQuery = function() {
-    // TODO - unhack this.
-    $scope.inputModel.query = document.getElementById("query-input").value;
-    $location.search("query", $scope.inputModel.query)
-    $location.search("renderType", $scope.inputModel.renderType)
-    $location.search("profile", $scope.inputModel.profile.toString())
+// A counter is an object with .inc() and .dec() methods,
+// as well as .pos() and .zero() predicates.
+function Counter() {
+  var count = 0;
+  this.inc = function() {
+    count++;
   };
-
-  $scope.$on("$locationChangeSuccess", function() {
-    // this triggers at least once (in the beginning).
-    var queries = $location.search();
-    $scope.inputModel.query = queries["query"] || "";
-    $scope.inputModel.renderType = queries["renderType"] || "line";
-    $scope.inputModel.profile = parseBool(queries["profile"]);
-    if ($scope.inputModel.query) {
-      launchRequest({
-        profile: $scope.inputModel.profile,
-        query:   $scope.inputModel.query
-      });
-    }
-  });
-
-  // true if the output should be tabular.
-  $scope.isTabular = function() {
-    return ["describe all", "describe metrics", "describe"].indexOf($scope.queryResult.name) >= 0;
+  this.dec = function() {
+    count--;
   };
-  $scope.hasProfiling = function() {
-    return !!($scope.queryResult && $scope.queryResult.profile);
+  this.pos = function() {
+    return count > 0;
+  };
+  this.zero = function() {
+    return count === 0;
   };
+};
+// A singleton Counter for launched queries.
+module.factory("$launchedQueries", function() {
+  return new Counter();
+});
+// A singleton Counter for waiting charts
+module.factory("$chartWaiting", function() {
+  return new Counter();
+});
 
-  $scope.screenState = function() {
-    if ($scope.launchedQuery > 0) {
-      return "loading";
-    } else if ($scope.launchedQuery === 0 && $scope.chartWaiting > 0) {
-      return "rendering";
-    } else if ($scope.queryResult && !$scope.queryResult.success) {
-      return "error";
-    } else {
-      return "rendered";
+// A ticket booth will give you a ticket with .next()
+// The ticket will be .valid() until another ticket has been asked for.
+function TicketBooth() {
+  var count = 0;
+  function Ticket(n) {
+    this.valid = function() {
+      return n === count;
     }
+  }
+  this.next = function() {
+    return new Ticket(++count);
   };
+}
+// A singleton ticketbooth for query counting
+module.factory("$queryTicketBooth", function() {
+  return new TicketBooth();
+});
 
-  $scope.chartWaiting = 0;
-  $scope.launchedQuery = 0;
+// chart-related functions
+function clearChart(chart) {
+  if (chart.chart !== null) {
+    chart.chart.clearChart();
+    chart.chart = null;
+  }
+};
+
+module.factory("$asyncRender", function($chartWaiting) {
+  return function($scope, chart, dataTable, options) {
```

Comment:
should just inject `$rootScope` and run `$rootScope.$apply` - that's a bit simpler.